File: CountLeaves.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Oct  6 23:29:53 2020

@author: kevin
"""

"""
Given a Binary Tree of size N , return the number of leaves in it.
"""
class btNode():
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None
        
def countLeaves(node):
    #If the left and right children are none, then the node is a leaf
    if(not node):
        return 0
    if(not node.left and not node.right):
        return 1
    return countLeaves(node.left) + countLeaves(node.right)

root = btNode(4)
root.left = btNode(2)
root.left.left = btNode(1)
root.left.right = btNode(3)
root.right = btNode(6)
root.right.left = btNode(5)

"""
          4
         /  \
        2    6
       /\    / 
      1  3  5
"""

print(countLeaves(root))
#3

File: CountLeaves.java
/* Given a Binary Tree of size N , You have to count leaves in it. */

class CountLeaves{
   public static void main(String[] args){
      TreeNode root = new TreeNode(1);
      root.left = new TreeNode(2);
      root.right = new TreeNode(3);
      root.left.left = new TreeNode(4);
      root.left.right = new TreeNode(5);
      root.right.left = new TreeNode(6);
      root.right.right = new TreeNode(7);
      System.out.println(root);
      /*          1
                /   \
               2     3
             /  \   /  \
            4    5 6    7
            
      */
      
      System.out.println(countLeaves(root)); //4
   }
   
   
   /**
   * Counts the number of leaves in a given binary tree
   *
   * @param root The root of a binary tree
   * @return An integer containing the number of leaves
   */
   static int countLeaves(TreeNode root){
      if(root.left == null && root.right == null){
         return 1;
      }
      int result = 0;
      if(root.left != null){
         result += countLeaves(root.left);
      }    
      if(root.right != null){
         result += countLeaves(root.right);
      }
      return result;
   }
}

class TreeNode{
   int data;
   TreeNode left;
   TreeNode right;
   
   public TreeNode(int data){
      this.data = data;
   }
   
   public String toString(){
      return data + "(" + left + ", " + right + ")";
   }
}